app/services/update_total_price.rb

```diff
@@ -14,21 +14,24 @@ def self.call(*args)
   end
 
   def call
-    if delivery
-      order.increment!(:total_price, @delivery.price)
-    end
-
-    if coupon
-      order.decrement!(:total_price, @order.discount)
-    end
-
-    if order_item && action
-      difference = order_item.price * order_item.quantity - order.discount(order_item.price)
-
-      case action
-      when :remove then order.decrement!(:total_price, difference)
-      when :add then order.increment!(:total_price, difference)
-      end
-    end
+    return order.increment!(:total_price, difference) if add_price?
+
+    order.decrement!(:total_price, difference) if minus_price?
+  end
+
+  def difference
+    return delivery.price if delivery
+
+    return order.discount if coupon
+
+    order_item.price * order_item.quantity - order.discount(order_item.price)
+  end
+
+  def add_price?
+    action == :add || delivery
+  end
+
+  def minus_price?
+    action == :remove || coupon
```

Comment:
These methods too hard to understand, also, you can pass action with coupon and delivery
`UpdateTotalPrice.call(current_order, action: :add, delivery: some_delivery)`
`UpdateTotalPrice.call(current_order, action: :remove, coupon: some_coupon)`